### ISSUE TITLE
writing to an SFTP file a large blob fails with a ERROR_SSH_INVALID_PACKET_SIZE

### ISSUE THREAD

**Issue by arthurdarcet** [ronf/asyncssh#164]
```python
async with asyncssh.connect(...) as conn:
    async with conn.start_sftp_client() as sftp:
        async with sftp.open('somewhere', 'w') as f:
            f.write('0' * 40000)
```

will fail, with three distinct issues:
 1. the info logger shows that a "exit status 11" packet is received (which is the invalid packet size error)
 2. the following traceback is logged because a None exception is used in `_cleanup`, which tries to set it as the exception for a waiter (None is not a valid exception)
```
Traceback (most recent call last):
  File \"/usr/local/lib/python3.6/site-packages/asyncssh/sftp.py\", line 790, in recv_packets
    packet = yield from self.recv_packet()
  File \"/usr/local/lib/python3.6/site-packages/asyncssh/sftp.py\", line 778, in recv_packet
    pktlen = yield from self._reader.readexactly(4)
  File \"/usr/local/lib/python3.6/asyncio/coroutines.py\", line 215, in coro
    res = yield from res
  File \"/usr/local/lib/python3.6/site-packages/asyncssh/stream.py\", line 470, in read
    raise asyncio.IncompleteReadError(buf, len(buf) + n)
asyncio.streams.IncompleteReadError: 0 bytes read on a total of 4 expected bytes

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/usr/local/lib/python3.6/site-packages/asyncssh/connection.py\", line 414, in _reap_task
    task.result()
  File \"/usr/local/lib/python3.6/site-packages/asyncssh/sftp.py\", line 801, in recv_packets
    yield from self._cleanup(None)
  File \"/usr/local/lib/python3.6/site-packages/asyncssh/sftp.py\", line 831, in _cleanup
    waiter.set_exception(exc)
TypeError: invalid exception object
```
 3. The upload actually hangs indefinitely because no exception / … is forwarded to the waiter. The script has to be killed manually

A workaround for this is to write small blobs to the remote file, to stay bellow the 32k per packet SSH limit

**Comment by ronf** [comment]
Regarding the use of None in _cleanup(), that is used in multiple places in AsyncSSH as an indication of an underlying connection close (following the way connection_lost typically works in Python's asyncio where either None or an exception can be passed). However, it looks like _cleanup() here is calling set_exception() without properly taking the possibility of None being passed into account.

**Comment by ronf** [comment]
You might want to try the following patch to see if it addresses the exception-related issue you saw:

```patch
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 924ff73..7927b65 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -798,7 +798,8 @@ class SFTPHandler(SSHPacketLogger):
         except PacketDecodeError as exc:
             yield from self._cleanup(SFTPError(FX_BAD_MESSAGE, str(exc)))
         except EOFError:
-            yield from self._cleanup(None)
+            yield from self._cleanup(SFTPError(FX_CONNECTION_LOST,
+                                               'Connection lost'))
         except (OSError, Error) as exc:
             yield from self._cleanup(exc)
 
```

There's really no reason I need to be passing in None in this case.

This won't address the root cause of the connection drop, but if you can provide some additional debug logging, I can try and help you get to the bottom of that.

Thanks for the report!

**Comment by arthurdarcet** [comment]
2 and 3. On the cleanup(None), yes, your patch ensures an exception is raised and the script doesn't hang, so I think that would be a good addition, thanks.

1. I didn't have time to get the exact limit yesterday, and assumed it was the default ssh limit, but you are right, the error only occurs if I try to write more than 256k at once, not 32k like I first reported - sorry for the messy issue… (the actual exact limit after which it breaks is `256 * 1024 - 25`)

I guess it would be nice to move the block_size handling from `_SFTPFileCopier` to `SFTPClientFile` so that the open/read/write interface could have the `block_size` argument (and more importantly, could have a sane default value for block size). Would you like a PR doing this?

**Comment by arthurdarcet** [comment]
and I forgot the logs if you still need them:
```
2018-08-29 16:03:03 |            asyncssh             |  DEBUG   | [conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
2018-08-29 16:03:03 |            asyncssh             |   INFO   | [conn=0, chan=0] Requesting new SSH session
2018-08-29 16:03:03 |            asyncssh             |  DEBUG   | [conn=0] Received debug message: Port forwarding disabled.
2018-08-29 16:03:03 |            asyncssh             |  DEBUG   | [conn=0] Received debug message: X11 forwarding disabled.
2018-08-29 16:03:03 |            asyncssh             |  DEBUG   | [conn=0] Received debug message: Pty allocation disabled.
2018-08-29 16:03:03 |            asyncssh             |  DEBUG   | [conn=0] Received debug message: Port forwarding disabled.
2018-08-29 16:03:03 |            asyncssh             |  DEBUG   | [conn=0] Received debug message: X11 forwarding disabled.
2018-08-29 16:03:03 |            asyncssh             |  DEBUG   | [conn=0] Received debug message: Pty allocation disabled.
2018-08-29 16:03:03 |            asyncssh             |   INFO   | [conn=0, chan=0]   Subsystem: sftp
2018-08-29 16:03:03 |          asyncssh.sftp          |   INFO   | [conn=0, chan=0] Starting SFTP client
2018-08-29 16:03:03 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Sending init, version=3
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Received version=3, extensions:
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0]   posix-rename@openssh.com: 1
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0]   statvfs@openssh.com: 2
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0]   fstatvfs@openssh.com: 2
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0]   hardlink@openssh.com: 1
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Sending realpath for /home/cli/simon_scp/upload/Glose
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Received 1 name
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0]   filename: /home/cli/simon_scp/upload/Glose, longname: /home/cli/simon_scp/upload/Glose
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Sending open for /home/cli/simon_scp/upload/Glose/.test, mode 0x1a
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Received handle 00000000
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Sending write for 262120 bytes at offset 0 in handle 00000000
2018-08-29 16:03:04 |            asyncssh             |   INFO   | [conn=0, chan=0] Received exit status 11
2018-08-29 16:03:04 |            asyncssh             |   INFO   | [conn=0, chan=0] Received channel close
2018-08-29 16:03:04 |          asyncssh.sftp          |   INFO   | [conn=0, chan=0] SFTP client exited: SFTP Error: Connection lost
2018-08-29 16:03:04 |            asyncssh             |   INFO   | [conn=0, chan=0] Closing channel
2018-08-29 16:03:04 |            asyncssh             |   INFO   | [conn=0, chan=0] Channel closed
2018-08-29 16:03:04 |          asyncssh.sftp          |  DEBUG   | [conn=0, chan=0] Sending close for handle 00000000
2018-08-29 16:03:04 |            asyncssh             |   INFO   | [conn=0] Closing connection
2018-08-29 16:03:04 |            asyncssh             |   INFO   | [conn=0] Sending disconnect: Disconnected by application (11)
2018-08-29 16:03:04 |            asyncssh             |   INFO   | [conn=0] Connection closed
2018-08-29 16:03:04 |           wyl.command           |  ERROR   | Uncaught SFTPError('SFTP Error: Connection not open')
Traceback (most recent call last):
  File "/Users/arthur/dev/dragonstone/dragonstone/apps/offers/commands/sas/__init__.py", line 44, in tmp
    await f.write('0' * (256 * 1024 - 24))
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 1480, in write
    yield from self._handler.write(self._handle, offset, data)
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 1098, in write
    UInt64(offset), String(data)))
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 877, in _make_request
    resptype, resp = yield from waiter
asyncssh.sftp.SFTPError: SFTP Error: Connection lost

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/arthur/dev/wyl/wyl/command_.py", line 47, in run
    res = loop.run_until_complete(task)
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 568, in run_until_complete
    return future.result()
  File "/Users/arthur/dev/dragonstone/dragonstone/apps/offers/commands/sas/__init__.py", line 44, in tmp
    await f.write('0' * (256 * 1024 - 24))
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/misc.py", line 170, in __aexit__
    yield from self._result.__aexit__(*exc_info)
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 1354, in __aexit__
    yield from self.close()
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 1691, in close
    yield from self._handler.close(self._handle)
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 1069, in close
    return (yield from self._make_request(FXP_CLOSE, String(handle)))
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 876, in _make_request
    self._send_request(pkttype, *args, waiter=waiter)
  File "/Users/arthur/dev/lib/asyncssh/asyncssh/sftp.py", line 856, in _send_request
    raise SFTPError(FX_NO_CONNECTION, 'Connection not open')
asyncssh.sftp.SFTPError: SFTP Error: Connection not open
```

(if the verbosity is too low, let me know how I should increase it)

**Comment by arthurdarcet** [comment]
I wouldn't consider the SFTPClientFile API to specify that only one request is eventually done, just like a call to the builtin `open().write` might result in several chunks being written separately to the disk.
I agree that there would be no way for the caller to now which chunk failed when an exception occurs, but I think that's to be expected: someone who wants more control would need to do the splitting himself of course.

Our use-case seems pretty basic: we need to write a csv file that we generate in-memory to an SFTP server. I really would prefer not having to write the data to a NamedTemporaryFile, then using the high-level helpers to upload it, but then - if i'm not mistaken - the only way to write data directly to the remote server is to use the `SFTPClientFile` api, which means we have to handle the `block_size` ourselves.
Imho, `SFTPClientFile` is a very nice API, and the lib would gain to expose it more. It mimics the builtin `open` in a very nice way and is very simple to use - except for that block_size…

I'm closing this as the issue here is resolved, if you change your mind on moving the block_size to `SFTPClientFile`, let me know, I'll be happy to help ; otherwise we'll work with what we have of course :).

Thanks for your help!

**Comment by ronf** [comment]
Thanks, Arthur. I've gone ahead and checked the exception handling improvements into the "develop" branch in e65d912.

I thought a bit more about your suggestion around SFTPClientFIle, and agree that the case of doing a large write from memory is compelling, and there's really no way to use the existing get/put/copy functions to do that given their current structure without creating temporary files, and that's definitely not good.

Taking a closer look at the code, I see there is one place today where multiple I/O operations are triggered by a single call, and that's when reading with an indeterminate size. In that case, the reads are done serially, not in parallel, and this only applies when no length is provided to read(), but it does exist. Also, if I were to make this change, I could provide arguments that would force the old single-request behavior, so perhaps that's not so bad.

I also discovered something interesting about calling read() with a really large length. When OpenSSH is the server, it actually won't return more than 64 KB. Even if there's more data than that at the requested offset, it silently reduces the length requested to 64 KB and returns only up to that amount. So, a caller would actually need to do multiple read calls if they wanted to get the full amount of data they asked for! This seems like a good reason to consider breaking the call into multiple requests, and there might even be some speed-up when doing that if read requests were done in parallel at different file offsets like I'm doing today on the file copies.

I also thought more about the concerns around issuing parallel requests and realized that it wouldn't actually be introducing any constraints on the caller that weren't already there, which is basically to avoid parallel requests with overlapping byte ranges if you want deterministic results. The issue around not being able to know what range(s) actually failed does apply, but even with a single write() call to the server, it's possible some data may have been successfully written before an error was returned and there's no way for a server to report that back.

So, long story short, I'm going to look into this. I appreciate your offer to help, and will definitely take you up on that from a testing point of view once I have something, or let you know if I run into problems.

**Comment by arthurdarcet** [comment]
That's great to hear, thank you for your detailed response!
We'll be happy to test once this has been implemented

**Comment by ronf** [comment]
Ok - this change is now available in the develop branch in commit c3187cd. I also made a small tweak to the previous fix on the exception handling in 2d9f110.

Parallel requests will be issued by default whenever a read or write of greater than 16 KB is issued. The behavior can be adjusted using new block_size and max_requests arguments on the open() call. Setting block_size to None will go back to issuing only a single request of whatever size you ask for (which may fail due to server limitations). You can also limit the number of parallel requests with max_requests, including setting it to 1 if for some reason you want the requests to be issued serially.

I also added a note in the documentation about the 256 KB message size limit in OpenSSH, as well as its 64 KB maximum read response size. This new block_size argument should be kept smaller than these values to avoid problems.
